torchnlp/metrics/rouge.py

```diff
@@ -0,0 +1,332 @@
+    """
+    calculate the n-grams.
+    Args:
+        n: n-gram to calculate
+        text: An array of tokens
+        exclusive: if True, the datastructure is set, else is set.
+    Returns:
+        A set of n-grams
+    """
+    ngram_set = Ngrams(exclusive=exclusive)
+    if type(text) == str:
+        text = text.split()
+    text_length = len(text)
+    index_ngram_end = text_length - n
+    for i in range(index_ngram_end + 1):
+        ngram_set.add(tuple(text[i:i + n]))
+    return ngram_set
+
+
+def _frp_rouge_n(eval_count, ref_count, overlapping_count):
+    """
+    compute f score, precision score and recall score.
+    Args:
+        eval_count: the evaluation sentence n-gram count
+        ref_count: the reference sentence n-gram count
+        overlapping_count: the overlapping n-gram between evaluation and reference.
+    """
+    if eval_count == 0:
+        precision = 0.0
+    else:
+        precision = overlapping_count / eval_count
+
+    if ref_count == 0:
+        recall = 0.0
+    else:
+        recall = overlapping_count / ref_count
+
+    f1_score = 2.0 * ((precision * recall) / (precision + recall + 1e-8))
+
+    return {"f": f1_score, "p": precision, "r": recall}
+
+
+def _frp_rouge_l(llcs, m, n, beta=None):
+    """
+    Computes the LCS-based F-measure score, Precision score and recall score.
+    Args:
+        llcs: Length of LCS
+        m: number of words in reference summary
+        n: number of words in candidate summary
+        beta: beta = P_lcs / R_lcs when ∂ F_lcs / ∂ R_lcs = ∂ F_lcs / ∂ P_lcs.
+            In DUC, beta is set to a very big number, and only R_lcs is considered.
+    Returns:
+        dictionary. 'f' for F-measure score, 'p' for Precision score and 'r' for recall score.
+    """
+    r_lcs = llcs / m
+    p_lcs = llcs / n
+    if not beta:
+        beta = p_lcs / (r_lcs + 1e-12)
+    num = (1 + (beta**2)) * r_lcs * p_lcs
+    denom = r_lcs + ((beta**2) * p_lcs)
+    f_lcs = num / (denom + 1e-12)
+    return {"f": f_lcs, "p": p_lcs, "r": r_lcs}
+
+
+def check_inverse(f, inv_f, eps=1e-5):
+    """
+    Check if function inv_f is the inverse function of function f
+    Args:
+        f: a function
+        inv_f: inverse function of f
+        eps: error threshold
+    Returns:
+        bool, if inv_f is the inverse function of f
+    """
+    for i in range(100000):
+        x = random.random() * 100000
+        y = f(x)
+        _x = inv_f(y)
+        if abs(x - _x) > eps:
+            return False
+    return True
+
+
+def check_increase(f):
+    """
+    Check if function f satisfies f(x + y) > f(x) + f(y)
+    Args:
+        f: function
+    Returns:
+        bool, if f(x + y) > f(x) + f(y) or not
+    """
+    for i in range(100000):
+        x = random.random() * 100000
+        y = random.random() * 100000
+        if f(x) + f(y) <= f(x + y):
+            return False
+    return True
+
+
+def _frp_rouge_w(wlcs, m, n, f=lambda x: x**2, inv_f=lambda x: math.sqrt(x),
```

Comment:
Do you need to repeat these lambdas a second time? I'm thinking about the principles of DRY.